### PR TITLE
Add more robust hostsfile parsing

### LINKED ISSUE #5063: Add more robust hostsfile parsing
Hostsfile support has been added as part of #3219. Two TODOs have also been added:

http://lxr.mozilla.org/servo/source/components/net/resource_task.rs#42
http://lxr.mozilla.org/servo/source/components/net/resource_task.rs#400

These relate to bad file path and contents. 

Additionally, other conditions should be checked for:
* comments
* mutliple hostnames for a single address

### PR DESCRIPTION
**Pull request by gilles-leblanc**
Adds hostsfile parsing support for:
- Tabs
- Comments (line and end of line)
- IPv4 address validation
- Basic IPv6 address validation
- End of line whitespaces
- Host name alias (multiple host names per address)

Fixes #5063

### CODE REVIEW

**Comment by jdm** [comment]
Nice work! I apologize for the delay in reviewing this.

**Comment by jdm** [comment]
Hmm. Try running `./mach update-cargo -p regex_macros --precise 0.1.8` and seeing if that results in changes to the Cargo.lock files in ports/cef and ports/gonk?

**Comment by gilles-leblanc** [comment]
@jdm Thanks.

Running the commande gives me:
```
$ ./mach update-cargo -p regex_macros --precise 0.1.8
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
no matching package named `regex` found (required by `net`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: = 0.1.14
ports/cef
package id specification `regex_macros` matched no packages
ports/gonk
package id specification `regex_macros` matched no packages
```

It didn't update the files. The libraries along exists in those version on crates.io. 
I'll investigate what is happening.

**Comment by jdm** [comment]
Maybe try `./mach update-carp -p regex --precise 0.1.8` instead? I'm not entirely sure of the relationship between the two packages.

**Comment by gilles-leblanc** [comment]
**tldr : I forced push a new update which fixes the problem but required some manual editing of 2 .lock files.** 

This was a fun problem to debug.

I played around with the cargo update and found out that component/util already referenced regex 0.1.14 and when I did a cargo update since I was referencing 0.1.8 in component/net that created a conflict and prevented the command to run.

If I changed the version for 0.1.14 I could now correctly update the cargo lock file for both ports. The problem was that it then automatically updated the dependency of regex_macros to 0.1.11 which doesn't compile with the version of Rust we are using (it needs a newer version). 

I would need to use an older version of regex_macros. So I tried updating the Cargo.lock file to demand 0.1.8 (which was the version I locked automatically when I first did it) but it didn't work when I specified a version and with no version specified it fell back to 0.1.11.

I got it to work by:
* Specifying the version numbers of regex and regex_macros in both net/Cargo.toml and util/Cargo.toml
* Manually updating the .lock files for cef and gonk to use regex_macros 0.1.8

Now I can build both servo and cef with no errors (gonk needs some unrelated env var).

I did a fetch upstream, merge master, rebase on branch along the way to check if that would update my version of Rust so I had to force push back the PR.

**Comment by jdm** [comment]
If it builds, I'll take it!
